include/seqan3/search/detail/policy_max_error.hpp

```diff
@@ -24,44 +25,90 @@ namespace seqan3::detail
+            if constexpr (search_traits_t::has_max_error_deletion)
+                errors.deletion = get_error_count(get<search_cfg::max_error_deletion>(cfg).value, query_size);
+
+            errors.total = std::min<uint32_t>(255, errors.substitution + errors.insertion + errors.deletion);
         }
+        else
+        { // set all error individually, if they have been set by the user
+            if constexpr (search_traits_t::has_max_error_total)
+                errors.total = get_error_count(get<search_cfg::max_error_total>(cfg).value, query_size);
+            if constexpr (search_traits_t::has_max_error_substitution)
+                errors.substitution = get_error_count(get<search_cfg::max_error_substitution>(cfg).value, query_size);
+            if constexpr (search_traits_t::has_max_error_insertion)
+                errors.insertion = get_error_count(get<search_cfg::max_error_insertion>(cfg).value, query_size);
+            if constexpr (search_traits_t::has_max_error_deletion)
+                errors.deletion = get_error_count(get<search_cfg::max_error_deletion>(cfg).value, query_size);
+        }
+
+        return errors;
+    }
+
+private:
 
-        return max_error;
+    /*!\brief Returns a uint8_t object directly given by the search_cfg::error_count or calculated by the
+     *        search_cfg::error_rate depending on the given variant.
+     *
+     * \tparam std::variant<search_cfg::error_count, search_cfg::error_rate> The error type.
+     * \tparam size_t                                                        The query size type.
+     *
+     * \param[in] error_variant The error saved in a variant, given by a uint8_t or double.
+     * \param[in] query_size    Query size, to calculate the uint8_t out of a given error rate. (maybe_unused)
+     *
+     * \throws std::invalid_argument if error_rate is not between 0 and 1.
+     *
+     * \details If seqan3::search_cfg::error_count is set in the configuration, its value already is of type uint8_t and
+     *          can be returned as is.
+     *          If seqan3::search_cfg::error_rate is set in the configuration, the error rates are converted to error
+     *          counts based on the length of the query sequence.
+     */
+    uint8_t get_error_count(std::variant<search_cfg::error_count, search_cfg::error_rate> const & error_variant,
+                            [[maybe_unused]] size_t query_size)
+    {
+        return std::visit([&query_size] (auto error)
+            {
+                if constexpr (std::same_as<decltype(error), search_cfg::error_count>)
+                    return error.get();
+                else
+                {
+                    // check correct error rate values.
+                    if (0.0 > error.get()  || error.get() > 1.0)
+                        throw std::invalid_argument("Error rates must be between 0 and 1.");
+                    return static_cast<uint8_t>(error.get() * query_size);
```

Comment:
You should do bounds checking.
If `error_rate = 0.5` and `query_size = 1000`, it does not fit in the total_errors anymore.
For the example it would return 244 (wrap around) instead of the (maybe) expected 255.

https://en.cppreference.com/w/cpp/algorithm/clamp can be useful for this